src/ImuSensor.cc

```diff
@@ -116,6 +116,18 @@ bool ImuSensor::Load(const sdf::Sensor &_sdf)
     return false;
   }
 
+  // Set orientation reference frame
+  // TODO(adityapande-1995) : Add support for named frames like
+  // ENU using ign-gazebo
+  if (_sdf.ImuSensor()->Localization() == "CUSTOM") {
+    if (_sdf.ImuSensor()->CustomRpyParentFrame() == "") {
+      this->SetOrientationReference(ignition::math::Quaterniond(
+        _sdf.ImuSensor()->CustomRpy()));
+    }
+  } else {
+    ignerr << "This tag is not yet supported" << std::endl;
```

Comment:
I'm not sure an error message is needed here, since ign-gazebo will handle any other types of `<localization>`.

You may consider a warning message if `CustomRpyParentFrame()` is not an empty string, since we don't currently support that use case.

Reply:
Done !

Reply:
thanks; I just pushed https://github.com/ignitionrobotics/ign-sensors/pull/178/commits/4848461a7db8c73aa37f010c46b3cba7a0bb96d7 to update the verbosity level used by the test so that the `ignwarn` message will show in the console output

Reply:
> since ign-gazebo will handle any other types of <localization>

@scpeters , how did you envision this being implemented? 

The IMU system in `ign-gazebo` should have knowledge of the current spherical coordinates, and it also should have access to `sdf::Imu::Localization`, so I guess it could call `sensors::ImuSensor::SetOrientationReference` as needed.

An alternative approach, if we want to keep most of the frame-conversion logic in `ign-sensors`, would be to expose an API like `sensors::ImuSensor::SetWorldFrameOrientation` so that the Gazebo system just lets the sensor know what the world is using, and the sensor performs all the calculations. I like this option better because it keeps all the SDF parsing within the sensor class.

Also worth noting that we don't support world frames other than ENU right now, so we should be able to default to that for a long time.

Reply:
I've added an example world with multiple IMU sensors with different `Localization` values in https://github.com/ignitionrobotics/ign-gazebo/commit/051f777dac5baff282c984c44ba0757f5964d6ad

> The IMU system in `ign-gazebo` should have knowledge of the current spherical coordinates, and it also should have access to `sdf::Imu::Localization`, so I guess it could call `sensors::ImuSensor::SetOrientationReference` as needed.
> 
> An alternative approach, if we want to keep most of the frame-conversion logic in `ign-sensors`, would be to expose an API like `sensors::ImuSensor::SetWorldFrameOrientation` so that the Gazebo system just lets the sensor know what the world is using, and the sensor performs all the calculations. I like this option better because it keeps all the SDF parsing within the sensor class.

can you describe the data types that you envision that `sensors::ImuSensor::SetWorldFrameOrientation` would use? I'm not sure exactly what you are proposing

Reply:
> can you describe the data types

I was thinking of an enum, but I guess it could also just be the strings that match the SDF spec. The enum is less error-prone though.

Reply:
ign-gazebo holds the SphericalCoordinates object, which can compute the rotation offset between the world frame and a frame like ENU. So are you proposing that we add an API to ImuSensor lets you specify a rotation offset between the world and the frame specified in an enum?

`sensors::ImuSensor::SetWorldFrameOrientation(const Quaterniond &_rot, const WorldFrameEnumType _relativeTo = ENU);`

then the IMU sensor would be responsible for mapping between the WorldFrameEnumType values? It's easy enough to map between ENU, NED, NWU.

Reply:
I think that `SphericalCoordinates` could be extended to support other tangent planes in the future, but at the moment it only supports ENU, which it calls [GLOBAL](https://github.com/ignitionrobotics/ign-math/blob/1c4a9f7e19e06ee6ea7566b5ae590d19c10c5a4b/include/ignition/math/SphericalCoordinates.hh#L59-L60) (`LOCAL` only allows changing the heading). There's no support for NED, NWU, etc.

I think that `ImuSensor` could use `sdf::Imu::Localization` to determine which "custom RPY" to apply, as [described on the SDF spec](https://github.com/ignitionrobotics/sdformat/blob/6404afdf959982f759526dea8700037b50fa7758/sdf/1.9/imu.sdf#L7-L30). So for example:

```
if localization == "NED"
  customRPY = this->SetOrientationReference(<rotation from NED to ENU>)
else if localization == "NWU"
  customRPY = this->SetOrientationReference(<rotation from NWU to ENU>)
...
```

Since for now Gazebo doesn't support frames other than ENU (if you ignore heading), I think that change here in `ign-sensors` would be enough. In the future, if we have support for other frames, then we could pass that information from Gazebo to `ign-sensors` like this for example:

```
/// \brief Set the tangential frame which the world is using. Defaults to ENU if not set.
sensors::ImuSensor::SetWorldFrameOrientation(WorldFrameEnumType _worldOrientation);
```

Then the logic inside `ImuSensor` becomes

```
if localization == "NED"
  customRPY = this->SetOrientationReference(<rotation from NED to worldOrientation>)
else if localization == "NWU"
  customRPY = this->SetOrientationReference(<rotation from NWU to worldOrientation>)
...
```

Reply:
I was under the impression that a heading offset could be applied to the world frame using spherical coordinates. In that case, the world frame would not be ENU. Does ign-gazebo not support heading offsets in the spherical coordinates for the world frame?

Reply:
> if you ignore heading

I guess this is the crux of the issue. Why would we ignore heading?

Reply:
> I guess this is the crux of the issue. Why would we ignore heading?

Ok, so the heading is at the same time too limited (because it doesn't allow flipping up and down) and too broad (because it accepts any arbitrary angle). This rules out turning the world into NED, but it's true that in theory we could have NWU, WSU and SEU besides ENU. It also allows confusing things like (NE)(NW)U if someone really wants to, or a 20° offset from North for example. So I guess I've been preferring to pretend it doesn't exist because it really complicates things :innocent: 

I suppose if we want to support custom headings of 90 degree increments, we could give them names (ENU / NWU / WSU / SEU) and implement the 2nd option I proposed above?

Reply:
if you are modeling a grid of buildings or roads that aren't aligned with cardinal directions (like downtown Mountain View), it could be convenient to use the heading angle to align the world X and Y axes with the axes of alignment used by the objects in the model

it's true that we should have made pitch/roll/up/down configurable as well, but I still see some value in arbitrary heading angles

Reply:
humm ok I see there's a use case there. Then I guess the heading needs to be passed to the IMU as well

Reply:
in that case does `sensors::ImuSensor::SetWorldFrameOrientation(const Quaterniond &_rot, const WorldFrameEnumType _relativeTo = ENU);` seem like a reasonable API?

Reply:
So in that case, `ign-gazebo` would populate `_rot` with just the heading?

Reply:
yeah, it could call `SetWorldFrameOrientation(Quaterniond(0, 0, heading), ENU);`

Reply:
Cool, that sounds reasonable to me :+1: